### DISCUSSION TITLE
Is Augmentation Fixed Or Different for Each Epoch? (3.0.0)

### DISCUSSION THREAD

**Discussion by jricheimer** [explosion/spaCy#7048] [Help: Coding & Implementations]
I'm using spacy 3.0 for text categorization. I am trying to add custom augmentation to the data (and I also have a custom reader, so I have some flexibility in how to use them).

However, from looking at the code in `training/loop.py`, it seems like the dataset is only generated once from the reader, then it is looped through again and again for each training epoch, without the data ever being regenerated by the reader. This is obviously nice from a speed perspective, but the downside is that if there is any randomness in the augmentation (which there often is), it will only be applied once and then fixed throughout training.

A naive workaround to this is to duplicate the dataset in advance by applying augmentations and saving the data, but besides for this being wasteful of space, it does not provide the theoretically limitless size of the dataset with continual random augmentation applied.

My question is: Have I understood what is happening in the code correctly? And if so, can we fix this easily?

**Comment by adrianeboyd** [comment]
I think you've understood the current behavior correctly. Off the top of my head, I would have expected you to be able to do this with an infinite stream for your train corpus, but it looks like this option has been removed so that the data is always shuffled (see fe298fa5).

You'd have to write a bunch of custom code around the training loop outside of `spacy train` in order to have this behavior, so I think we should look into this again...
